core/domain/html_cleaner.py

```diff
@@ -136,11 +136,12 @@ def get_rte_components(html_string):
             customization_args = {}
             for attr in oppia_custom_tag_attrs[tag_name]:
                 # Unescape special HTML characters such as '&quot;'.
-                attr_val = parser.unescape(component_tag[attr])
-                # Adds escapes so that things like '\frac' aren't
-                # interpreted as special characters.
-                attr_val = attr_val.encode('unicode_escape')
-                customization_args[attr] = json.loads(attr_val)
+                if attr in str(component_tag):
```

Comment:
Why we need this if?

Reply:
Because in some cases in the `oppia-noninteractive` tag, the attribute `caption-with-value`  is optional and it gives an error in that case. So this is to check if it is there in the component_tag or not.

Reply:
Please explain how you came to this conclusion (that the attribute is optional)?

If it truly is optional we need to model it differently: each component should have a list of optional and required attributes and we should check each separately. But ideally all attributes are required.

@bansalnitish did you encounter this when working with the RTE?

Reply:
There is no caption-with-value attribute in one of the sample explorations, [Lazy Magician](https://github.com/oppia/oppia/blob/develop/data/explorations/binary_search/The%20Lazy%20Magician.yaml#L258).
Also, in the `assets/rich_text_components_definitions.js` [link](https://github.com/oppia/oppia/blob/develop/assets/rich_text_components_definitions.js#L62) it is written that it is optional.

I too found the flaw when I was testing a demo exploration (with exploration id '2') and the `html_cleaner`'s get_rte_components gave an error. This might not have been caught because the function isn't used anywhere in the code and the tests don't include `oppia-noninteractive-image` tag without the `caption-with-value`.

Reply:
Thanks for the explanation. "Optional" in the docstring means that it could be the empty string, not that the field doesn't exist at all.

I think we may need @bansalnitish to weigh in here, since he's recently been introspecting all RTE components on the server and has a good idea of what is going on there. Also /cc @AllanYangZhou in case he can help.

Reply:
Oh! Got it. Thanks! But we do have demo explorations that contain the tag without the caption-with-value.

Reply:
Hi @seanlip,

@ishucr7 is in constant touch with me, we are discussing all the problems he is facing (on hangouts). Though I'm not following this PR, if need will do that too. 
Happy to help :).

Thanks

Reply:
Thanks! It seems to me like @ishucr7 is seeing some issues with the caption tag on images possibly being absent. I was wondering if you had noticed anything like this when you were investigating the RTE stuff.

Also, apparently demo exploration 2 is invalid in that respect; it might be a good idea to figure out why. Does it have an exploration schema version that was older than when captions were first introduced? If so, that might be an explanation.

Reply:
The schema version of the demo exploration 2 (`The Lazy Magician` in  `data/explortaions/binary_search`) is 17. But I wasn't able to find when were the captions introduced. Hi @bansalnitish , Could you help me out ?

Reply:
Also the states schema version when these changes were introduced was 7 (https://github.com/oppia/oppia/pull/2215/files#diff-c8a6978dd86c922ed0cc864523778fe3L69) and the schema version was 10 (https://github.com/oppia/oppia/blob/54e5a4956aa86e5a302d2bb26866c60c6487b3f0/core/domain/exp_domain.py#L2237).

Reply:
Ah OK. So it seems to me like that caption field was introduced without an exploration migration. In general that's not correct.

We should fix this by adding a caption attribute with the empty string to all image components if one does not already exist. I will file an issue for this. If you depend on this functionality then we should fix this first rather than use an "if" branch here; it should be reasonably straightforward.

Reply:
Hi @seanlip  Since this is migration and the explorations will change after the release, I think its better to keep the if statement till the migration is done. Because the tests for the getting the filenames from the exploration uses this function and the demo exploration no 2.
Once the migration is done, I'll remove the if condition.
What do you say?
Also, shall I create an issue for this and mention it there to be done after the July's release ?

Reply:
No -- we shouldn't have indeterminism in the codebase; the fix for indeterminism is to make things more deterministic.

I suggest we get @bansalnitish's PR in (#5179) and then we can assume that every exploration has captions. At that point I think your tests will pass because, when the lazy magician exploration is loaded, it would be automatically upgraded to the latest schema version.

Reply:
Yes, will remove it then.